x11rb/src/rust_connection/stream.rs

```diff
@@ -351,33 +363,39 @@ fn do_write(
     bufs: &[IoSlice<'_>],
     fds: &mut Vec<RawFdContainer>,
 ) -> Result<usize> {
-    use nix::sys::socket::{sendmsg, ControlMessage, MsgFlags, SockaddrLike};
+    use rustix::fd::{AsFd, BorrowedFd};
+    use rustix::io::Errno;
+    use rustix::net::{sendmsg_noaddr, SendAncillaryBuffer, SendAncillaryMessage, SendFlags};
 
-    fn sendmsg_wrapper<S: SockaddrLike>(
-        fd: RawFd,
+    fn sendmsg_wrapper(
+        fd: BorrowedFd<'_>,
         iov: &[IoSlice<'_>],
-        cmsgs: &[ControlMessage<'_>],
-        flags: MsgFlags,
-        addr: Option<&S>,
+        cmsgs: &mut SendAncillaryBuffer<'_, '_, '_>,
+        flags: SendFlags,
     ) -> Result<usize> {
         loop {
-            match sendmsg(fd, iov, cmsgs, flags, addr) {
+            match sendmsg_noaddr(fd, iov, cmsgs, flags) {
                 Ok(n) => return Ok(n),
                 // try again
-                Err(nix::Error::EINTR) => {}
+                Err(Errno::INTR) => {}
                 Err(e) => return Err(e.into()),
             }
         }
     }
 
-    let fd = stream.as_raw_fd();
+    let fd = stream.as_fd();
 
     let res = if !fds.is_empty() {
-        let fds = fds.iter().map(|fd| fd.as_raw_fd()).collect::<Vec<_>>();
-        let cmsgs = [ControlMessage::ScmRights(&fds[..])];
-        sendmsg_wrapper::<()>(fd, bufs, &cmsgs, MsgFlags::empty(), None)?
+        let fds = fds.iter().map(|fd| fd.as_fd()).collect::<Vec<_>>();
+        let rights = SendAncillaryMessage::ScmRights(&fds);
+
+        let mut cmsg_space = vec![0u8; rights.size()];
+        let mut cmsg_buffer = SendAncillaryBuffer::new(&mut cmsg_space);
+        assert!(cmsg_buffer.push(rights));
```

Comment:
Does this really work? The code provides spaces for `num_fds` bytes, but shouldn't this need at least `num_fds * sizeof(int)`?

Edit: After reading a bit further, the receiving code uses `cmsg_space!(ScmRights(MAX_FDS_RECEIVED)`. Couldn't / shouldn't this use something similar, but with `num_fds` instead of `MAX_FDS_RECEIVED`? (End-Edit)

Random idea that I am not sure about: `man 7 unix` contains:

>  The kernel constant SCM_MAX_FD defines a limit on the number of file descriptors in the array.  [...] SCM_MAX_FD has the value 253 (or 255 before Linux 2.6.38).

This could allocate 1024 bytes as an array on the stack. Something like `// surely this is always gonna be enough (tm)`. ;-)
Or perhaps better: A comment pointing to `SCM_MAX_FD` being at most 255 according to `man 7 unix` and then using `255 * 4` for the size.

Edit: Oh and please add some kind of message to the assert saying something like "unrealistic large number of FDs being passed", or so.

Reply:
Either I was wrong back then or the code changed enough that things now work. This uses `SendAncillaryMessage::size` and that function does the `CMSG_SPACE` stuff.